zstd.go

```diff
@@ -20,35 +17,64 @@ var zstdDecMap sync.Map
 
 var zstdAvailableEncoders sync.Map
 
+var zstdCheckedOutEncoders int
+var zstdMutex = &sync.Mutex{}
+var zstdEncoderReturned = sync.NewCond(zstdMutex)
+var zstdTestingDisableConcurrencyLimit bool
+
 func getZstdEncoderChannel(params ZstdEncoderParams) chan *zstd.Encoder {
 	if c, ok := zstdAvailableEncoders.Load(params); ok {
 		return c.(chan *zstd.Encoder)
 	}
-	c, _ := zstdAvailableEncoders.LoadOrStore(params, make(chan *zstd.Encoder, zstdMaxBufferedEncoders))
+	limit := runtime.GOMAXPROCS(0)
+	c, _ := zstdAvailableEncoders.LoadOrStore(params, make(chan *zstd.Encoder, limit))
 	return c.(chan *zstd.Encoder)
 }
 
+func newZstdEncoder(params ZstdEncoderParams) *zstd.Encoder {
+	encoderLevel := zstd.SpeedDefault
+	if params.Level != CompressionLevelDefault {
+		encoderLevel = zstd.EncoderLevelFromZstd(params.Level)
+	}
+	zstdEnc, _ := zstd.NewWriter(nil, zstd.WithZeroFrames(true),
+		zstd.WithEncoderLevel(encoderLevel),
+		zstd.WithEncoderConcurrency(1))
+	return zstdEnc
+}
+
 func getZstdEncoder(params ZstdEncoderParams) *zstd.Encoder {
+	zstdMutex.Lock()
+	defer zstdMutex.Unlock()
+
+	limit := runtime.GOMAXPROCS(0)
+	for zstdCheckedOutEncoders >= limit && !zstdTestingDisableConcurrencyLimit {
```

Comment:
1. Is `zstdCheckedOutEncoders` and the capacity of of zstdEncoderChannel are doing the same kind of control?
2. Is seems hard to follow if the `limit` value from line 49 are different from line 52 or line 29, might need some comment or explanation.

Reply:
Some process could have changed the `GOMAXPROCS` in between calls. The proper behavior after calling a `sync.Semaphore.Wait()` is to refresh all values before performing any tests.